OutOfSchool/OutOfSchool.WebApi/Controllers/ChatWorkshopController.cs

```diff
@@ -56,30 +56,179 @@ public class ChatWorkshopController : ControllerBase
+            => this.SetReadDatetimeInMessagesForCurrentUserAsync(id, this.IsProviderAChatRoomParticipantAsync);
+
+        private Task<bool> IsParentAChatRoomParticipantAsync(ChatRoomWorkshopDto chatRoom)
+        {
+            var userId = this.GetUserId();
+
+            return validationService.UserIsParentOwnerAsync(userId, chatRoom.ParentId);
+        }
+
+        private Task<bool> IsProviderAChatRoomParticipantAsync(ChatRoomWorkshopDto chatRoom)
+        {
+            var userId = this.GetUserId();
+
+            return validationService.UserIsWorkshopOwnerAsync(userId, chatRoom.WorkshopId);
+        }
+
+        private string GetUserId()
+        {
+            var userId = HttpContext.User.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Sub)
+                ?? throw new AuthenticationException($"Can not get user's claim {nameof(IdentityResourceClaimsTypes.Sub)} from HttpContext.");
+
+            return userId;
+        }
+
+        private Role GetUserRole()
+        {
+            var userRoleName = HttpContext.User.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Role)
+                ?? throw new AuthenticationException($"Can not get user's claim {nameof(IdentityResourceClaimsTypes.Sub)} from HttpContext.");
+
+            Role userRole = (Role)Enum.Parse(typeof(Role), userRoleName, true);
+
+            return userRole;
+        }
+
+        private void LogWarningAboutUsersTryingToGetNotOwnChatRoomInformation(long chatRoomId)
+        {
+            var messageToLog = $"{this.GetUserRole()} with UserId:{this.GetUserId()} is trying to get not his own chat room: {nameof(chatRoomId)}={chatRoomId}.";
+            logger.Warning(messageToLog);
+        }
+
+        private async Task<IActionResult> GetRoomByIdAsync(long chatRoomId, Func<ChatRoomWorkshopDto, Task<bool>> userHasRights)
         {
             try
             {
-                var chatRoom = await roomService.GetByIdAsync(id).ConfigureAwait(false);
+                var chatRoom = await roomService.GetByIdAsync(chatRoomId).ConfigureAwait(false);
 
-                if (!(chatRoom is null) &&
-                    await this.UserHasRigtsForChatRoomAsync(chatRoom.WorkshopId, chatRoom.ParentId).ConfigureAwait(false))
+                var isChatRoomValid = !(chatRoom is null) && await userHasRights(chatRoom).ConfigureAwait(false);
+
+                if (isChatRoomValid)
                 {
                     return Ok(chatRoom);
                 }
 
-                var messageToLog = $"{HttpContext.User.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Role)} with UserId:{HttpContext.User.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Sub)} is trying to get not his own chat room: {nameof(id)} {id}.";
-                logger.Warning(messageToLog);
+                this.LogWarningAboutUsersTryingToGetNotOwnChatRoomInformation(chatRoomId);
```

Comment:
so if with such id still does not exist warning about access will be logged ?

Reply:
No, if the chat room with such Id does not exist, or the user has no rights to it, we return NoContent. We don't tell the user about another, not his own existing chat rooms.
And we log that the user is trying to get forbidden to him chat room because if the room does not exist it is forbidden too.